packages/js/product-editor/src/components/details-name-block/edit.tsx

```diff
@@ -2,36 +2,175 @@
  * External dependencies
  */
 import { __ } from '@wordpress/i18n';
-import { createElement, createInterpolateElement } from '@wordpress/element';
-import { TextControl } from '@woocommerce/components';
+import {
+	createElement,
+	Fragment,
+	createInterpolateElement,
+	useState,
+} from '@wordpress/element';
+
 import { useBlockProps } from '@wordpress/block-editor';
+import { cleanForSlug } from '@wordpress/url';
+import { useSelect, useDispatch } from '@wordpress/data';
+import {
+	PRODUCTS_STORE_NAME,
+	WCDataSelector,
+	Product,
+} from '@woocommerce/data';
+import {
+	Button,
+	BaseControl,
+	// @ts-expect-error `__experimentalInputControl` does exist.
+	__experimentalInputControl as InputControl,
+} from '@wordpress/components';
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment
 // @ts-ignore No types for this exist yet.
 // eslint-disable-next-line @woocommerce/dependency-group
-import { useEntityProp } from '@wordpress/core-data';
+import { useEntityProp, useEntityId } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { AUTO_DRAFT_NAME } from '../../utils';
+import { EditProductLinkModal } from '../edit-product-link-modal';
+import { useValidation } from '../../hooks/use-validation';
+import { useProductHelper } from '../../hooks/use-product-helper';
 
 export function Edit() {
 	const blockProps = useBlockProps();
-	const [ name, setName ] = useEntityProp( 'postType', 'product', 'name' );
+
+	const { updateProductWithStatus } = useProductHelper();
+
+	const [ showProductLinkEditModal, setShowProductLinkEditModal ] =
+		useState( false );
+
+	const { editEntityRecord } = useDispatch( 'core' );
+	const productId = useEntityId( 'postType', 'product' );
+	const product: Product = useSelect( ( select ) =>
+		select( 'core' ).getEditedEntityRecord(
+			'postType',
+			'product',
+			productId
+		)
+	);
```

Comment:
Instead of returning the entire product object here, I would just return the items you need: `permalink`, 'slug` and `status`. As the `useSelect` function is memoized, if someone where to add images or a category to the product it would re-trigger the render of this block, which would be unnecessary.

Reply:
Hmm yeah you are right, I guess we could refactor it after.
Although I think you already removed most of the `product` use cases within the Modal component when introducing the `saveHandler` callback. If you move the `recordEvent` out into the callbacks as well the only use for product within the Modal component is the `slug`.

But this can be a follow up for sure!

Reply:
Agree with the above, and as mentioned this refactor will be a little easier/long-lasting if we do so after removing the previous iteration of the editor.